File: src/tasks/storage.ts
import { randomUUID } from "node:crypto";
import { mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { getProjectDir } from "../config/paths.ts";
import type { Task, TaskStatus } from "./types.ts";

/** Path to the tasks.json file for a project. */
export function getTasksPath(projectPath?: string): string {
	return join(getProjectDir(projectPath), "tasks.json");
}

/** Read and parse tasks.json. Returns [] if file doesn't exist. */
export async function loadTasks(projectPath?: string): Promise<Task[]> {
	const path = getTasksPath(projectPath);
	try {
		const raw = readFileSync(path, "utf-8");
		return JSON.parse(raw) as Task[];
	} catch {
		return [];
	}
}

/** Write tasks array to tasks.json (pretty-printed JSON). */
export async function saveTasks(tasks: Task[], projectPath?: string): Promise<void> {
	const path = getTasksPath(projectPath);
	const dir = join(path, "..");
	mkdirSync(dir, { recursive: true });
	writeFileSync(path, JSON.stringify(tasks, null, "\t"), "utf-8");
}

/** Create a new task with status "pending", save it, and return it. */
export async function createTask(
	title: string,
	sessionId: string,
	projectPath?: string,
	details?: string,
): Promise<Task> {
	const tasks = await loadTasks(projectPath);
	const now = new Date().toISOString();

	const task: Task = {
		id: randomUUID(),
		title,
		status: "pending",
		created: now,
		updated: now,
		session_id: sessionId,
		...(details !== undefined ? { details } : {}),
	};

	tasks.push(task);
	await saveTasks(tasks, projectPath);
	return task;
}

/** Update a task by id. Throws if not found. */
export async function updateTask(
	id: string,
	updates: Partial<Pick<Task, "status" | "title" | "details">>,
	projectPath?: string,
): Promise<Task> {
	const tasks = await loadTasks(projectPath);
	const index = tasks.findIndex((t) => t.id === id);

	if (index === -1) {
		throw new Error(`Task not found: ${id}`);
	}

	const task = tasks[index] as Task;
	if (updates.status !== undefined) task.status = updates.status;
	if (updates.title !== undefined) task.title = updates.title;
	if (updates.details !== undefined) task.details = updates.details;
	task.updated = new Date().toISOString();

	await saveTasks(tasks, projectPath);
	return task;
}

const STATUS_ORDER: TaskStatus[] = ["in_progress", "blocked", "pending", "done"];

/** Format tasks for injection into system prompt. Flags stale tasks from other sessions. */
export function formatTasksSummary(tasks: Task[], currentSessionId: string): string {
	if (tasks.length === 0) {
		return "No tasks tracked.";
	}

	const lines: string[] = [];

	for (const status of STATUS_ORDER) {
		const group = tasks.filter((t) => t.status === status);
		if (group.length === 0) continue;

		lines.push(`## ${status.replace("_", " ").toUpperCase()}`);
		for (const task of group) {
			const staleFlag = task.session_id !== currentSessionId ? " [stale]" : "";
			const detailsSuffix = task.details ? ` — ${task.details}` : "";
			lines.push(`- [${task.id}] ${task.title}${detailsSuffix}${staleFlag}`);
		}
		lines.push("");
	}

	return lines.join("\n");
}

File: src/tasks/storage.rs
//! Task persistence: read/write tasks.json under the project dir.

use std::path::PathBuf;

use anyhow::{anyhow, Result};
use chrono::Utc;
use uuid::Uuid;

use super::types::{Task, TaskStatus};
use crate::config::paths::get_project_dir;

pub fn get_tasks_path(project_path: Option<&str>) -> PathBuf {
    get_project_dir(project_path).join("tasks.json")
}

pub fn load_tasks(project_path: Option<&str>) -> Vec<Task> {
    let path = get_tasks_path(project_path);
    let content = match std::fs::read_to_string(&path) {
        Ok(c) => c,
        Err(_) => return Vec::new(),
    };
    serde_json::from_str(&content).unwrap_or_default()
}

pub fn save_tasks(tasks: &[Task], project_path: Option<&str>) -> Result<()> {
    let path = get_tasks_path(project_path);
    if let Some(parent) = path.parent() {
        std::fs::create_dir_all(parent)?;
    }
    let serialized = serde_json::to_string_pretty(tasks)?;
    std::fs::write(&path, serialized)?;
    Ok(())
}

pub fn create_task(
    title: &str,
    session_id: &str,
    project_path: Option<&str>,
    details: Option<&str>,
) -> Result<Task> {
    let mut tasks = load_tasks(project_path);
    let now = Utc::now().to_rfc3339();
    let task = Task {
        id: Uuid::new_v4().to_string(),
        title: title.to_string(),
        status: TaskStatus::Pending,
        created: now.clone(),
        updated: now,
        session_id: session_id.to_string(),
        details: details.map(String::from),
    };
    tasks.push(task.clone());
    save_tasks(&tasks, project_path)?;
    Ok(task)
}

#[derive(Debug, Clone, Default)]
pub struct TaskUpdate {
    pub status: Option<TaskStatus>,
    pub title: Option<String>,
    pub details: Option<String>,
}

pub fn update_task(id: &str, update: TaskUpdate, project_path: Option<&str>) -> Result<Task> {
    let mut tasks = load_tasks(project_path);
    let task = tasks
        .iter_mut()
        .find(|t| t.id == id)
        .ok_or_else(|| anyhow!("Task not found: {id}"))?;
    if let Some(s) = update.status {
        task.status = s;
    }
    if let Some(t) = update.title {
        task.title = t;
    }
    if let Some(d) = update.details {
        task.details = Some(d);
    }
    task.updated = Utc::now().to_rfc3339();
    let updated = task.clone();
    save_tasks(&tasks, project_path)?;
    Ok(updated)
}

const STATUS_ORDER: &[TaskStatus] = &[
    TaskStatus::InProgress,
    TaskStatus::Blocked,
    TaskStatus::Pending,
    TaskStatus::Done,
];

pub fn format_tasks_summary(tasks: &[Task], current_session_id: &str) -> String {
    if tasks.is_empty() {
        return "No tasks tracked.".to_string();
    }
    let mut lines = Vec::new();
    for status in STATUS_ORDER {
        let group: Vec<&Task> = tasks.iter().filter(|t| t.status == *status).collect();
        if group.is_empty() {
            continue;
        }
        lines.push(format!(
            "## {}",
            status.as_str().replace('_', " ").to_uppercase()
        ));
        for task in group {
            let stale = if task.session_id != current_session_id {
                " [stale]"
            } else {
                ""
            };
            let details = task
                .details
                .as_deref()
                .map(|d| format!(" — {d}"))
                .unwrap_or_default();
            lines.push(format!(
                "- [{}] {}{}{}",
                task.id, task.title, details, stale
            ));
        }
        lines.push(String::new());
    }
    lines.join("\n")
}
